validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py

```diff
@@ -0,0 +1,209 @@
+from validphys.closuretest.closure_checks import check_multifit_replicas
+
+from reportengine import collect
+
+
+@dataclasses.dataclass(frozen=True)
+class PCAInternalMulticlosureLoader:
+    """
+    Parameters
+    ----------
+    closures_th: list
+        list containing validphys.results.ThPredictionsResult objects
+        for each fit
+
+    law_th: ThPredictionsResult object
+        underlying law theory predictions
+
+    pc_basis: np.array
+        basis of principal components
+
+    n_comp: int
+        number of principal components kept after regularisation
+
+    covmat_pca: np.array
+        regularised covariance matrix computed from replicas
+        of theory predictions
+
+    sqrt_covmat_pca: np.array
+        cholesky decomposed covariance matrix
+    """
+
+    closures_th: list
+    law_th: ThPredictionsResult
+    pc_basis: np.array
+    n_comp: int
+    covmat_pca: np.array
+    sqrt_covmat_pca: np.array
+
+
+@check_multifit_replicas
+def internal_multiclosure_dataset_loader_pca(
+    internal_multiclosure_dataset_loader,
+    explained_variance_ratio=0.99,
+    _internal_max_reps=None,
+    _internal_min_reps=20,
+):
+    """
+    Similar to multiclosure.internal_multiclosure_dataset_loader but returns
+    PCA regularised covariance matrix, where the covariance matrix has been computed
+    from the replicas of the theory predictions.
+
+    Parameters
+    ----------
+    internal_multiclosure_dataset_loader: tuple
+        closure fits theory predictions,
+        underlying law theory predictions,
+        covariance matrix,
+        sqrt covariance matrix
+
+    explained_variance_ratio: float, default is 0.99
+
+    _internal_max_reps: int, default is None
+        Maximum number of replicas used in the fits
+        this is needed to check that the number of replicas is the same for all fits
+
+    _internal_min_reps: int, default is 20
+        Minimum number of replicas used in the fits
+        this is needed to check that the number of replicas is the same for all fits
+
+    Returns
+    -------
+    PCAInternalMulticlosureLoader
+    """
+    closures_th, law_th, _, _ = internal_multiclosure_dataset_loader
+
+    reps = np.asarray([th.error_members for th in closures_th])
+
+    # compute the covariance matrix of the theory predictions for each fit
+    _covmats = [np.cov(rep, rowvar=True, bias=True) for rep in reps]
+
+    # compute the mean covariance matrix
+    _covmat_mean = np.mean(_covmats, axis=0)
+
+    # diagonalize the mean covariance matrix and only keep the principal components
+    # that explain the required variance
+
+    if _covmat_mean.shape == ():
+        return PCAInternalMulticlosureLoader(
+            closures_th=closures_th,
+            law_th=law_th,
+            pc_basis=1,
+            n_comp=1,
+            covmat_pca=_covmat_mean,
+            sqrt_covmat_pca=np.sqrt(_covmat_mean),
+        )
+
+    eighvals, eigvecs = np.linalg.eigh(_covmat_mean)
```

Comment:
Maybe this block can be put in another function since I guess it could be reused

Reply:
Yes this makes sense to have perhaps. I added a function called `eigendecomposition` that returns the eigenvectors and eigenvalues orderered in descendent way